### PR TITLE
Text submission action request to Phoenix API

### PR DESCRIPTION
**Pull request by weerd**
# _PULL REQUEST OVERVIEW_

### What does this PR do?
This PR performs some connections between the front-end and the back-end. It deals with validation for the `TextSubmissionAction` component while setting up the `PostRequest` to be able to deal with validation for other actions in the future as well, setting the validation rules by using a `switch` based on the action `type`. 

It _only_ sends request to Phoenix API from the front-end, but nothing is going to Rogue yet. I need to cleanup the `store()` method on the `CampaignPostsController.php` file, since we'll need to start accounting for more than just a photo submission.

### Any background context you want to provide?
Sorry I went over the 10 files rule in this PR. I tried to whittle it down. Lots of changes happening @_@

### What are the relevant tickets/cards?
Refs [Pivotal ID#155866202](https://www.pivotaltracker.com/story/show/155866202)

### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Added appropriate feature/unit tests.

### CODE REVIEW

File: app/Http/Controllers/Api/CampaignPostsController.php [hunk 1]
```diff
@@ -55,39 +56,43 @@ public function index($id, Request $request)
...
-    public function store($id, Request $request)
+    public function store($id, PostRequest $request)
     {
-        $this->validate($request, [
-            'media' => 'required',  //@TODO: add file|image
-            'caption' => 'required|min:4|max:60',
-            'impact' => 'required|integer|min:1',
-            'whyParticipated' => 'required',
-        ]);
+        // @TODO: Reactivate this once in upcoming pull request:
```
Comment (weerd) [inline app/Http/Controllers/Api/CampaignPostsController.php]: Had to comment this out for now. Requires a bit of a fix and didn't need to include in this PR. Will be in upcoming PR!

File: app/Http/Requests/PostRequest.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace App\Http\Requests;
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class PostRequest extends FormRequest
```
Comment (mendelB) [inline app/Http/Requests/PostRequest.php]: Is the idea here for us to eventually be replacing the `StorePostRequest`with this one?

File: app/Http/Requests/PostRequest.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+{
+    /**
+     * Determine if the user is authorized to make this request.
+     *
+     * @return bool
+     */
+    public function authorize()
+    {
+        return true;
```
Comment (mendelB) [inline app/Http/Requests/PostRequest.php]: Assuming this is temp right?

File: app/Http/Controllers/CampaignController.php [hunk 4]
```diff
@@ -61,6 +61,7 @@ public function show($slug, Request $request)
                 'isAuthenticated' => auth()->check(),
                 'role' => auth()->user() ? auth()->user()->role : null,
                 'source' => $request->query('utm_source'),
+                'token' => auth()->user() ? auth()->user()->access_token : null,
```
Comment (mendelB) [inline app/Http/Controllers/CampaignController.php]: Cool! Did not realize we even had this available

File: resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js [hunk 5]
```diff
@@ -7,58 +9,84 @@ import Card from '../../Card';
 import './text-submission-action.scss';
 
 class TextSubmissionAction extends React.Component {
-  constructor(props) {
-    super(props);
+  state = {
```
Comment (mendelB) [inline resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js]: 😱 Are we all set to use this new class format?

File: resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js [hunk 6]
```diff
@@ -7,58 +9,84 @@ import Card from '../../Card';
...
+    formData.append('campaignId', this.props.campaignId);
+    formData.append('type', this.props.type);
+    formData.append('text', this.state.textValue);
+
+    if (this.props.legacyCampaignId && this.props.legacyCampaignRunId) {
+      formData.append('details', JSON.stringify({
+        legacyCampaignId: this.props.legacyCampaignId,
+        legacyCampaignRunId: this.props.legacyCampaignRunId,
+      }));
+    }
```
Comment (mendelB) [inline resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js]: I never completely understood this, these values are Ashes related, right? And Rogue needs them for some reason?

File: resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js [hunk 7]
```diff
@@ -7,58 +9,84 @@ import Card from '../../Card';
...
   className: PropTypes.string,
-  contentfulId: PropTypes.string.isRequired,
+  id: PropTypes.string.isRequired,
+  legacyCampaignId: PropTypes.string,
+  legacyCampaignRunId: PropTypes.string,
+  storeCampaignPost: PropTypes.func.isRequired,
   textFieldLabel: PropTypes.string,
   textFieldPlaceholder: PropTypes.string,
   title: PropTypes.string,
-  userId: PropTypes.string.isRequired,
```
Comment (mendelB) [inline resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js]: 🎊  No more `userId` due to access token?

File: resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js (hunk 6 above)
Comment (mendelB) [reply resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js]: Yes, Mendel. They are Ashes related, and we're conditionally passing them along to Rogue in a `details` JSON field as a nicety, so we have this data available.

File: resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js (hunk 5 above)
Comment (mendelB) [reply resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js]: Yes, Mendel, we are! Get excited! Just be sure to do an `npm install` before you get too jumpy.

File: app/Http/Requests/PostRequest.php (hunk 2 above)
Comment (mendelB) [reply app/Http/Requests/PostRequest.php]: Mmm hmm, how right you are! We are formalizing a `switch` to address all the various Post items we might be using right in this here `PostRequest`!

File: app/Http/Requests/PostRequest.php (hunk 3 above)
Comment (weerd) [reply app/Http/Requests/PostRequest.php]: Nope. I think you change it based on other scenarios. For the most part the request is typically allowed.

File: app/Http/Controllers/CampaignController.php (hunk 4 above)
Comment (weerd) [reply app/Http/Controllers/CampaignController.php]: I ganked this from a similar approach in Rogue 😉

File: resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js (hunk 7 above)
Comment (weerd) [reply resources/assets/components/actions/TextSubmissionAction/TextSubmissionAction.js]: Yep yep! Rogue grabs it in v3 endpoint from OAuth, so no need to pass it along!

**Comment by weerd** [comment]
I enjoyed your responses to your own questions based on our IRL convo, haha!

**Pull request merged**